### PR TITLE
New signal sprites system 🚦

### LINKED ISSUE #5172: Adapt how signals are displayed
The idea is to entrust the back end with the responsibility of managing signal sprites.

## Back

- [ ] Have an endpoints to list available signaling systems
- [ ] Have endpoints following map libre format to retrieve atlases

## Front

- [ ] Uses endpoints to load multiple atlases given the available signaling system

### PR DESCRIPTION
**Pull request by flomonster**
What is done:
- [X] Signals atlas is stored and handled by editoast
- [X] A config file specifies how match sprites with their parameters
- [x] Editoast is using the config file to generate its signal layer
- Editoast handle all signaling systems
  - [X] BAL
  - [x] TVM
  - [x] BAPR 
- [x] The front uses the new atlas
- [x] The front uses the new system in its map style

### CODE REVIEW

File: editoast/src/assets/signal_sprites/BAPR/sprites.json [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    "y": 256
+  },
+  "S VL": {
+    "height": 128,
+    "pixelRatio": 1,
+    "width": 128,
+    "x": 0,
+    "y": 384
+  }
+}
```
Comment (leovalais) [inline editoast/src/assets/signal_sprites/BAPR/sprites.json]: Missing newline

File: editoast/src/assets/signal_sprites/BAPR/sprites@3x.json [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    "y": 768
+  },
+  "S VL": {
+    "height": 384,
+    "pixelRatio": 3,
+    "width": 384,
+    "x": 768,
+    "y": 768
+  }
+}
```
Comment (leovalais) [inline editoast/src/assets/signal_sprites/BAPR/sprites@3x.json]: Missing newline

File: editoast/src/assets/signal_sprites/TVM/sprites.json [hunk 3]
```diff
@@ -0,0 +1,9 @@
+{
+  "REP TGV": {
+    "height": 128,
+    "pixelRatio": 1,
+    "width": 128,
+    "x": 0,
+    "y": 0
+  }
+}
```
Comment (leovalais) [inline editoast/src/assets/signal_sprites/TVM/sprites.json]: Missing newline

File: editoast/src/assets/signal_sprites/TVM/sprites@2x.json [hunk 4]
```diff
@@ -0,0 +1,9 @@
+{
+  "REP TGV": {
+    "height": 128,
+    "pixelRatio": 2,
+    "width": 128,
+    "x": 0,
+    "y": 0
+  }
+}
```
Comment (leovalais) [inline editoast/src/assets/signal_sprites/TVM/sprites@2x.json]: missing newline

File: editoast/src/assets/signal_sprites/TVM/sprites@3x.json [hunk 5]
```diff
@@ -0,0 +1,9 @@
+{
+  "REP TGV": {
+    "height": 192,
+    "pixelRatio": 3,
+    "width": 192,
+    "x": 0,
+    "y": 0
+  }
+}
```
Comment (leovalais) [inline editoast/src/assets/signal_sprites/TVM/sprites@3x.json]: missing newline

File: editoast/src/client/mod.rs [hunk 6]
```diff
@@ -176,3 +176,10 @@ pub fn get_root_url() -> String {
 pub fn get_app_version() -> Option<String> {
     env::var("OSRD_GIT_DESCRIBE").ok()
 }
+
+/// Retrieve the assets path
+pub fn get_assets_path() -> PathBuf {
+    env::var("ASSETS_PATH")
+        .unwrap_or(String::from("./src/assets"))
```
Comment (leovalais) [inline editoast/src/client/mod.rs]: It feels weird to have an `assets/` folder inside `src/`. Wouldn't `editoast/assets/` make more sense?

File: editoast/src/generated_data/signal.rs [hunk 7]
```diff
@@ -2,31 +2,98 @@ use async_trait::async_trait;
...
 use crate::error::Result;
 use crate::infra_cache::InfraCache;
-use crate::schema::ObjectType;
+use crate::schema::sprite_config::{SpriteConfig, SpriteConfigs};
+use crate::schema::{LogicalSignal, ObjectType};
 use crate::tables::infra_layer_signal::dsl;
 
 pub struct SignalLayer;
 
+/// Found the corresponding sprite id for a given logical signal
```
Comment (leovalais) [inline editoast/src/generated_data/signal.rs]:
```suggestion
/// Finds the corresponding sprite id for a given logical signal
```

File: editoast/src/generated_data/signal.rs [hunk 8]
```diff
@@ -2,31 +2,98 @@ use async_trait::async_trait;
...
+    infra_cache: &InfraCache,
+    updated_signals: T,
+) -> Result<()> {
+    let sprite_configs = SpriteConfig::load();
+    let mut group_by_sprite: HashMap<_, Vec<_>> = HashMap::new();
+
+    for signal_id in updated_signals {
+        let signal = match infra_cache.signals().get(signal_id) {
+            Some(signal) => signal,
+            None => continue,
```
Comment (leovalais) [inline editoast/src/generated_data/signal.rs]: Don't we want to try to fetch the signal in the DB in this case, and return an error if missing? Or at least `expect` if the cache is not supposed to be out of sync?

**Review by leovalais — COMMENTED** [review]
Tested the back only. Left a few, mostly cosmetic, comments.
Some CI tests fail. Thanks for adding the SVGs! 🚥

**Review by nicolaswurtz — APPROVED** [review]

File: editoast/src/assets/signal_sprites/BAPR/sprites.json (hunk 1 above)
Comment (flomonster) [reply editoast/src/assets/signal_sprites/BAPR/sprites.json]: These files are generated. I can adapt the script to add a linefeed.

File: editoast/src/generated_data/signal.rs (hunk 8 above)
Comment (flomonster) [reply editoast/src/generated_data/signal.rs]: I decided to expect it.

**Review by leovalais — APPROVED** [review]

**Pull request merged**